File: ideonomy/triangulate.py
"""Executable triangulation for questions with no computable oracle — P-9.

The top-ranked premier skill, made real. For irreducibly normative / aesthetic
/ strategic questions, you must not fabricate a scalar oracle and optimize it
with false rigor. Instead: dimensionalize the value space, gather >=2
*independent* judgments per axis, and report the spread. Where independent
judges agree, the axis is settled and compresses; where they disagree, the
disagreement is the signal — it localizes exactly where the question genuinely
underdetermines its answer. Contested axes are named as residue (kind
`contested_axis`) and can be fed straight into the cross-session ledger (P-10),
so P-9 and P-10 compose.

Organon: P5 dimensionalize + P21 judge-panel + M4 cross-model triangulation +
P37 residue-seed. The gate is honesty about the absence of a gate: you either
produce independent reads per axis, or you name the judgment irreducible and
its owner — never a single manufactured number.

Judges are injectable so the logic is testable offline; for real use, back them
with `CommandModel` (any model CLI). Stdlib-only.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any, Callable, Optional

# A judge scores one axis of one question.
Judge = Callable[[str, str], "Verdict"]


@dataclass
class Verdict:
    judge: str
    axis: str
    lean: float          # -1 (fails the axis) .. +1 (excels); 0 = neutral/uncertain
    stance: str          # one-line reason


@dataclass
class AxisResult:
    axis: str
    verdicts: list[Verdict] = field(default_factory=list)

    @property
    def leans(self) -> list[float]:
        return [v.lean for v in self.verdicts]

    @property
    def mean(self) -> float:
        xs = self.leans
        return round(sum(xs) / len(xs), 3) if xs else 0.0

    @property
    def spread(self) -> float:
        xs = self.leans
        return round(max(xs) - min(xs), 3) if xs else 0.0

    def contested(self, spread_threshold: float = 1.0) -> bool:
        """Contested if judges disagree in sign, one is uncertain while
        another is confident, or the spread is wide."""
        xs = self.leans
        if len(xs) < 2:
            return True                       # one read is not triangulation
        signs = {1 if x > 0.15 else (-1 if x < -0.15 else 0) for x in xs}
        if {1, -1} <= signs:                  # genuine sign disagreement
            return True
        if 0 in signs and any(abs(x) >= 0.6 for x in xs):
            return True                       # uncertain vs confident is a split
        return self.spread >= spread_threshold


@dataclass
class Triangulation:
    question: str
    axes: list[AxisResult] = field(default_factory=list)

    def contested(self) -> list[AxisResult]:
        return [a for a in self.axes if a.contested()]

    def settled(self) -> list[AxisResult]:
        return [a for a in self.axes if not a.contested()]

    def report(self) -> str:
        lines = [f"question: {self.question}",
                 f"axes: {len(self.axes)}  settled: {len(self.settled())}  "
                 f"contested: {len(self.contested())}", ""]
        for a in self.axes:
            tag = "CONTESTED" if a.contested() else "settled"
            lines.append(f"[{tag}] {a.axis}  mean={a.mean:+.2f} spread={a.spread:.2f}")
            for v in a.verdicts:
                lines.append(f"    {v.judge}: {v.lean:+.2f}  {v.stance}")
        if self.contested():
            lines += ["", "irreducible judgments (no oracle — owner must decide):"]
            for a in self.contested():
                lines.append(f"  - {a.axis}: judges split; owner=___ grounds=___")
        return "\n".join(lines)

    def to_residue(self, ledger: Any, origin: str = "P-9") -> list[str]:
        """Write each contested axis to a residue ledger (P-10). Returns ids."""
        ids = []
        for a in self.contested():
            r = ledger.add(f"contested axis: {a.axis} (spread {a.spread:.2f})",
                           "contested_axis", origin=origin)
            ids.append(r.id)
        return ids


def triangulate(question: str, axes: list[str], judges: list[Judge]) -> Triangulation:
    """Run every judge on every axis. The gate: >=2 independent judges, else
    this is not triangulation — raise and tell the caller to name the owner."""
    if len(judges) < 2:
        raise ValueError(
            "triangulation needs >=2 independent judges. With fewer, do not "
            "fabricate a verdict — name the judgment irreducible and its owner.")
    if not axes:
        raise ValueError("no value axes given; dimensionalize the question first "
                         "(P5) — a question with no named axes cannot be triangulated.")
    tri = Triangulation(question=question)
    for axis in axes:
        ar = AxisResult(axis=axis)
        for j in judges:
            ar.verdicts.append(j(axis, question))
        tri.axes.append(ar)
    return tri


# ----------------------------------------------------- model-backed judges
def model_judge(model: Callable[[str], str], name: str) -> Judge:
    """A judge backed by any model callable (e.g. CommandModel)."""
    def judge(axis: str, question: str) -> Verdict:
        prompt = (
            f"Judge ONE axis of the question below, independently. Do not try to "
            f"give an overall verdict.\n\nQuestion: {question}\nAxis: {axis}\n\n"
            "Reply with exactly two lines:\n"
            "LEAN: <a number from -1.0 (fails this axis) to +1.0 (excels), "
            "0 if genuinely uncertain>\n"
            "WHY: <one sentence>")
        reply = model(prompt)
        return Verdict(judge=name, axis=axis, lean=_parse_lean(reply),
                       stance=_parse_why(reply))
    return judge


def dimensionalize(question: str, model: Callable[[str], str], k: int = 4) -> list:
    """P5: ask a model for the 2-5 value axes the vague predicate comprises."""
    prompt = (
        f"The question below turns on a vague predicate (good/right/better). "
        f"Name the {k} concrete, independent value-axes it actually comprises — "
        f"the dimensions along which one would judge it. One per line, terse, "
        f"no numbering.\n\nQuestion: {question}")
    reply = model(prompt)
    axes = [ln.strip(" -*\t0123456789.)") for ln in reply.splitlines() if ln.strip()]
    axes = [a for a in axes if a]
    if not axes:
        # Refuse to manufacture the single vague axis this skill exists to
        # prevent. No axes -> the caller must dimensionalize by hand (P5).
        raise ValueError(
            "dimensionalization returned no axes; name the value axes "
            "yourself (P5) — refusing to fall back to a manufactured scalar.")
    return axes[:k]


def _parse_lean(text: str) -> float:
    m = re.search(r"LEAN:\s*([+-]?\d*\.?\d+)", text, re.IGNORECASE)
    if not m:
        m = re.search(r"([+-]?\d*\.?\d+)", text)
    try:
        return max(-1.0, min(1.0, float(m.group(1)))) if m else 0.0
    except ValueError:
        return 0.0


def _parse_why(text: str) -> str:
    m = re.search(r"WHY:\s*(.+)", text, re.IGNORECASE)
    if m:
        return m.group(1).strip()
    lines = text.strip().splitlines()
    return lines[-1][:160] if lines else "(no reason given)"


def main(argv: Optional[list] = None) -> int:
    import argparse
    from .models import CommandModel

    ap = argparse.ArgumentParser(
        prog="ideonomy.triangulate",
        description="Triangulate a no-oracle question: >=2 independent judges "
                    "per value axis; report the spread, never a single number.")
    ap.add_argument("question")
    ap.add_argument("--axis", action="append", default=[], help="value axis (repeatable)")
    ap.add_argument("--judge", action="append", default=[], required=True,
                    help="model command with {prompt} (repeatable; need >=2)")
    args = ap.parse_args(argv)
    if len(args.judge) < 2:
        ap.error("triangulation needs >=2 independent --judge commands; with "
                 "one judge, name the judgment irreducible and its owner instead")
    if len(set(args.judge)) < len(args.judge):
        ap.error("duplicate --judge commands: identical judges are one judge, "
                 "not a panel — independence is the point (M4). Vary the model "
                 "or the framing. (Distinct commands are necessary, not "
                 "sufficient; true independence is on you.)")

    judges_models = [CommandModel(c, name=f"judge-{i+1}") for i, c in enumerate(args.judge)]
    axes = args.axis or dimensionalize(args.question, judges_models[0])
    judges = [model_judge(m, m.name) for m in judges_models]
    tri = triangulate(args.question, axes, judges)
    print(tri.report())
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/Ideonomy/Triangulate.hs
-- | Executable triangulation for questions with no computable oracle — P-9.
--
-- The top-ranked premier skill, made real. For irreducibly normative,
-- aesthetic, or strategic questions, you must not fabricate a scalar oracle
-- and optimize it with false rigor. Instead: dimensionalize the value space,
-- gather >=2 /independent/ judgments per axis, and report the spread. Where
-- independent judges agree, the axis is settled and compresses; where they
-- disagree, the disagreement is the signal — it localizes exactly where the
-- question genuinely underdetermines its answer. Contested axes are named as
-- residue (kind @contested_axis@) and can be fed straight into the
-- cross-session ledger (P-10), so P-9 and P-10 compose.
--
-- Organon: P5 dimensionalize + P21 judge-panel + M4 cross-model triangulation
-- + P37 residue-seed. The gate is honesty about the absence of a gate: you
-- either produce independent reads per axis, or you name the judgment
-- irreducible and its owner — never a single manufactured number.
--
-- Judges are injectable so the logic is testable offline; for real use, back
-- them with 'command' (any model CLI): @ideonomy triangulate@.
module Ideonomy.Triangulate
  ( Judge, Verdict (..), AxisResult (..), Triangulation (..)
  , leans, mean, spread, contested, contestedAt, contestedAxes, settledAxes
  , report, residueItems, triangulate, modelJudge, dimensionalize
  , parseLean, parseWhy, cli
  ) where

import Data.Char (isSpace)
import Data.List (tails)
import Ideonomy.Cli (opts, parseArgs, positionals, usage)
import Ideonomy.Models (Model, clamp, command, lastLabeled, leadingNumber, mkCommand)
import Ideonomy.Util (joinWith, nub', roundTo, showFixed, showSigned, strip, stripCI, truncateTo)

-- | A judge scores one axis of one question: @judge axis question@.
type Judge = String -> String -> IO Verdict

data Verdict = Verdict
  { judge :: String
  , axis :: String
  , lean :: Double     -- ^ -1 (fails the axis) .. +1 (excels); 0 = neutral/uncertain
  , stance :: String   -- ^ one-line reason
  } deriving (Eq, Show)

data AxisResult = AxisResult { axis :: String, verdicts :: [Verdict] }
  deriving (Eq, Show)

data Triangulation = Triangulation { question :: String, axes :: [AxisResult] }
  deriving (Eq, Show)

leans :: AxisResult -> [Double]
leans a = [v.lean | v <- a.verdicts]

mean :: AxisResult -> Double
mean a = case leans a of
  [] -> 0
  xs -> roundTo 3 (sum xs / fromIntegral (length xs))

spread :: AxisResult -> Double
spread a = case leans a of
  [] -> 0
  xs -> roundTo 3 (maximum xs - minimum xs)

-- | Contested if judges disagree in sign, one is uncertain while another is
-- confident, or the spread is wide (threshold 1.0).
contested :: AxisResult -> Bool
contested = contestedAt 1.0

contestedAt :: Double -> AxisResult -> Bool
contestedAt threshold a
  | length xs < 2 = True                                 -- one read is not triangulation
  | 1 `elem` signs && (-1) `elem` signs = True           -- genuine sign disagreement
  | 0 `elem` signs && any ((>= 0.6) . abs) xs = True     -- uncertain vs confident is a split
  | otherwise = spread a >= threshold
  where
    xs = leans a
    signs = [if x > 0.15 then 1 else if x < -0.15 then -1 else 0 :: Int | x <- xs]

contestedAxes :: Triangulation -> [AxisResult]
contestedAxes t = filter contested t.axes

settledAxes :: Triangulation -> [AxisResult]
settledAxes t = filter (not . contested) t.axes

report :: Triangulation -> String
report t = joinWith "\n" $
  [ "question: " ++ t.question
  , "axes: " ++ show (length t.axes) ++ "  settled: " ++ show (length (settledAxes t))
      ++ "  contested: " ++ show (length (contestedAxes t))
  , "" ]
  ++ concat
    [ ("[" ++ tag ++ "] " ++ a.axis ++ "  mean=" ++ showSigned 2 (mean a) ++ " spread=" ++ showFixed 2 (spread a))
      : ["    " ++ v.judge ++ ": " ++ showSigned 2 v.lean ++ "  " ++ v.stance | v <- a.verdicts]
    | a <- t.axes, let tag = if contested a then "CONTESTED" else "settled" ]
  ++ (if null (contestedAxes t) then [] else
        ["", "irreducible judgments (no oracle — owner must decide):"]
        ++ ["  - " ++ a.axis ++ ": judges split; owner=___ grounds=___" | a <- contestedAxes t])

-- | Each contested axis as residue for the ledger (P-10): @(text, kind)@,
-- kind always @contested_axis@.
residueItems :: Triangulation -> [(String, String)]
residueItems t =
  [ ("contested axis: " ++ a.axis ++ " (spread " ++ showFixed 2 (spread a) ++ ")", "contested_axis")
  | a <- contestedAxes t ]

-- | Run every judge on every axis. The gate: >=2 independent judges, else
-- this is not triangulation — refuse and tell the caller to name the owner.
triangulate :: String -> [String] -> [Judge] -> IO (Either String Triangulation)
triangulate question axes judges
  | length judges < 2 = pure (Left
      "triangulation needs >=2 independent judges. With fewer, do not fabricate a verdict — name the judgment irreducible and its owner.")
  | null axes = pure (Left
      "no value axes given; dimensionalize the question first (P5) — a question with no named axes cannot be triangulated.")
  | otherwise = do
      rs <- mapM (\ax -> AxisResult ax <$> mapM (\j -> j ax question) judges) axes
      pure (Right (Triangulation question rs))

-- ----------------------------------------------------- model-backed judges

-- | A judge backed by any model (e.g. 'command').
modelJudge :: Model -> String -> Judge
modelJudge model nm axis question = do
  reply <- model prompt
  pure Verdict { judge = nm, axis = axis, lean = parseLean reply, stance = parseWhy reply }
  where
    prompt =
      "Judge ONE axis of the question below, independently. Do not try to "
      ++ "give an overall verdict.\n\nQuestion: " ++ question ++ "\nAxis: " ++ axis ++ "\n\n"
      ++ "Reply with exactly two lines:\n"
      ++ "LEAN: <a number from -1.0 (fails this axis) to +1.0 (excels), "
      ++ "0 if genuinely uncertain>\n"
      ++ "WHY: <one sentence>"

-- | P5: ask a model for the 2-5 value axes the vague predicate comprises.
-- No axes -> refuse to manufacture the single vague axis this skill exists
-- to prevent; the caller must dimensionalize by hand.
dimensionalize :: String -> Model -> Int -> IO (Either String [String])
dimensionalize question model k = do
  reply <- model prompt
  let axes = [a | ln <- lines reply, not (null (strip ln)), let a = trimSet ln, not (null a)]
  pure $ if null axes
    then Left "dimensionalization returned no axes; name the value axes yourself (P5) — refusing to fall back to a manufactured scalar."
    else Right (take k axes)
  where
    prompt =
      "The question below turns on a vague predicate (good/right/better). "
      ++ "Name the " ++ show k ++ " concrete, independent value-axes it actually comprises — "
      ++ "the dimensions along which one would judge it. One per line, terse, "
      ++ "no numbering.\n\nQuestion: " ++ question
    junk = " -*\t0123456789.)"
    trimSet = dropWhile (`elem` junk) . reverse . dropWhile (`elem` junk) . reverse

-- | The LAST @LEAN: <number>@; when no labeled number exists, the first
-- number anywhere in the text; 0 when there is none. Clamped to [-1, 1].
parseLean :: String -> Double
parseLean text = case labeled "LEAN:" text of
  [] -> maybe 0 clamp (firstNumber text)
  xs -> clamp (last xs)

-- | The LAST @WHY:@ line; else the last line (truncated), else a placeholder.
parseWhy :: String -> String
parseWhy text = case lastLabeled "WHY:" text of
  Just w | not (null w) -> w
  _ -> case lines (strip text) of
    [] -> "(no reason given)"
    ls -> truncateTo 160 (last ls)

labeled :: String -> String -> [Double]
labeled label text =
  [n | tl <- tails text, Just rest <- [stripCI label tl], Just n <- [leadingNumber (dropWhile isSpace rest)]]

firstNumber :: String -> Maybe Double
firstNumber text = case [n | tl <- tails text, Just n <- [leadingNumber tl]] of
  [] -> Nothing
  (n : _) -> Just n

-- --------------------------------------------------------------------- CLI

usageLine :: String
usageLine = "usage: ideonomy triangulate QUESTION --judge CMD --judge CMD [--judge CMD]... [--axis AXIS]..."

-- | @ideonomy triangulate QUESTION --judge CMD --judge CMD ... [--axis AXIS]...@
-- Judges are model commands with @{prompt}@; without @--axis@ the first judge
-- dimensionalizes the question.
cli :: [String] -> IO ()
cli argv = do
  let a = parseArgs [] argv
  case (positionals a, opts "judge" a) of
    ([question], cmds@(c1 : _ : _))
      | length (nub' cmds) < length cmds -> usage (usageLine ++ "\nerror: "
          ++ "duplicate --judge commands: identical judges are one judge, not a panel — "
          ++ "independence is the point (M4). Vary the model or the framing. "
          ++ "(Distinct commands are necessary, not sufficient; true independence is on you.)")
      | otherwise -> do
          let models = [("judge-" ++ show i, command (mkCommand c ("judge-" ++ show i))) | (i, c) <- zip [1 :: Int ..] cmds]
          axes <- case opts "axis" a of
            [] -> dimensionalize question (command (mkCommand c1 "judge-1")) 4 >>= either (ioError . userError) pure
            xs -> pure xs
          tri <- triangulate question axes [modelJudge m nm | (nm, m) <- models] >>= either (ioError . userError) pure
          putStrLn (report tri)
    ([_], [_]) -> usage (usageLine ++ "\nerror: triangulation needs >=2 independent --judge commands; "
                         ++ "with one judge, name the judgment irreducible and its owner instead")
    _ -> usage usageLine
